eng/packages/http-client-csharp/generator/Azure.Generator/src/Providers/ResourceProvider.cs

```diff
@@ -0,0 +1,345 @@
+                var operationDeclaration = Declare("operation", New.Instance(typeof(ArmOperation), _clientDiagonosticsField, This.Property("Pipeline"), _restClientField.Invoke(requestMethod.Signature.Name, PopulateArguments(requestMethod.Signature.Parameters)), responseVariable, Static(typeof(OperationFinalStateVia)).Property(((OperationFinalStateVia)operation.LongRunning!.FinalStateVia).ToString())), out var operationVariable);
+                tryStatement.Add(operationDeclaration);
+                tryStatement.Add(new IfStatement(KnownAzureTypes.WaitUntil.Equal(Static(typeof(WaitUntil)).Property(nameof(WaitUntil.Completed))))
+                {
+                    isAsync
+                    ? operationVariable.Invoke("WaitForCompletionAsync", [KnownParameters.CancellationTokenParameter], null, isAsync).Terminate()
+                    : operationVariable.Invoke("WaitForCompletion", KnownParameters.CancellationTokenParameter).Terminate()
+                });
+                tryStatement.Add(Return(operationVariable));
+            }
+            else
+            {
+                tryStatement.Add(new IfStatement(responseVariable.Property("Value").Equal(Null))
+            {
+                ((KeywordExpression)ThrowExpression(New.Instance(typeof(RequestFailedException), responseVariable.Invoke("GetRawResponse")))).Terminate()
+            });
+                tryStatement.Add(Return(Static(typeof(Response)).Invoke(nameof(Response.FromValue), New.Instance(Type, This.Property("Client"), responseVariable.Property("Value")), responseVariable.Invoke("GetRawResponse"))));
+            }
+            return tryStatement;
+        }
+
+        private static CSharpType GetResponseType(MethodProvider convenienceMethod, bool isAsync) => isAsync ? convenienceMethod.Signature.ReturnType?.Arguments[0]! : convenienceMethod.Signature.ReturnType!;
+
+        private ValueExpression[] PopulateArguments(IReadOnlyList<ParameterProvider> parameters)
+        {
+            var arguments = new List<ValueExpression>();
+            foreach (var parameter in parameters)
+            {
+                if (parameter.Name.Equals("subscriptionId", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    arguments.Add(Static(typeof(Guid)).Invoke(nameof(Guid.Parse), This.Property(nameof(ArmResource.Id)).Property(nameof(ResourceIdentifier.SubscriptionId))));
+                    continue;
+                }
+                if (parameter.Name.Equals("resourceGroupName", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    arguments.Add(This.Property(nameof(ArmResource.Id)).Property(nameof(ResourceIdentifier.ResourceGroupName)));
+                    continue;
+                }
+
+                // TODO: handle parent
+                if (parameter.Name.Equals(_contextualParameters.Last(), StringComparison.InvariantCultureIgnoreCase))
```

Comment:
is it correct that we are getting the last item by `Last()` while `_contextualParameters` is actually a HashSet?

Looks like when we are using this list, we are not really relying on the fact that "it is a set", we only make it a set because we want to reduplicate. If so, we could change it to a readonlylist and here we could do `_contextualParameters[^1]`

Reply:
yeah, it should be list instead of set